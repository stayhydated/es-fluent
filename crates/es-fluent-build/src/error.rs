use thiserror::Error;

#[derive(Debug, Error)]
pub enum FluentBuildError {
    #[error(
        "No i18n.toml configuration file found at {0}. Please create one before using this tool."
    )]
    NoI18nConfig(std::path::PathBuf),

    #[error("Failed to parse i18n.toml configuration: {0}")]
    ConfigParseError(#[from] es_fluent_toml::I18nConfigError),

    #[error("Cannot create i18n output directory: {0}")]
    CreateDirError(#[from] std::io::Error),

    #[error(transparent)]
    FluentParserError(#[from] es_fluent_sc_parser::error::FluentScParserError),

    #[error(transparent)]
    FluentGenerateError(#[from] es_fluent_generate::error::FluentGenerateError),
}
