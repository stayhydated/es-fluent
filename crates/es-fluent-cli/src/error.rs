use thiserror::Error;

/// An error that can occur in the CLI.
#[derive(Debug, Error)]
pub enum CliError {
    /// An error that occurs when parsing the `i18n.toml` configuration file.
    #[error("Failed to parse i18n.toml configuration: {0}")]
    ConfigParse(String),

    /// An error that occurs when creating the i18n output directory.
    #[error("Cannot create i18n output directory: {0}")]
    CreateDir(#[from] std::io::Error),

    /// An error that occurs when discovering the workspace metadata.
    #[error("Failed to discover workspace metadata: {0}")]
    WorkspaceDiscovery(String),

    /// An error that occurs when watching files.
    #[error("File watching error: {0}")]
    Watch(#[from] notify::Error),

    /// An error that occurs when parsing Fluent files.
    #[error("Fluent parser error: {0}")]
    FluentParser(String),

    /// An error that occurs when generating Fluent files.
    #[error("Fluent generation error: {0}")]
    FluentGenerate(String),

    /// An internal application error.
    #[error("An internal application error occurred: {0}")]
    Internal(String),
}

impl From<es_fluent_toml::I18nConfigError> for CliError {
    fn from(err: es_fluent_toml::I18nConfigError) -> Self {
        CliError::ConfigParse(err.to_string())
    }
}

impl From<cargo_metadata::Error> for CliError {
    fn from(err: cargo_metadata::Error) -> Self {
        CliError::WorkspaceDiscovery(err.to_string())
    }
}

impl From<es_fluent_sc_parser::error::FluentScParserError> for CliError {
    fn from(err: es_fluent_sc_parser::error::FluentScParserError) -> Self {
        CliError::FluentParser(err.to_string())
    }
}

impl From<es_fluent_generate::error::FluentGenerateError> for CliError {
    fn from(err: es_fluent_generate::error::FluentGenerateError) -> Self {
        CliError::FluentGenerate(err.to_string())
    }
}
