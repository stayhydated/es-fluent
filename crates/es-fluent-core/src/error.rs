use proc_macro_error2::{abort, abort_call_site, emit_error};
use proc_macro2::Span;

#[derive(Debug, thiserror::Error)]
pub enum EsFluentCoreError {
    /// Error related to Fluent attribute parsing
    #[error("Attribute error: {message}")]
    AttributeError { message: String, span: Option<Span> },

    /// Error related to variant consistency
    #[error("Variant '{variant_name}' error: {message}")]
    VariantError {
        message: String,
        variant_name: String,
        span: Option<Span>,
    },

    /// Error related to field processing
    #[error("{}", field_error_fmt(.message, .field_name))]
    FieldError {
        message: String,
        field_name: Option<String>,
        span: Option<Span>,
    },

    /// Error with conversions or transformations
    #[error("Transform error: {message}")]
    TransformError { message: String, span: Option<Span> },
}

fn field_error_fmt(message: &str, field_name: &Option<String>) -> String {
    match field_name {
        Some(name) => format!("Field '{}' error: {}", name, message),
        None => format!("Field error: {}", message),
    }
}

impl EsFluentCoreError {
    pub fn span(&self) -> Option<Span> {
        match self {
            EsFluentCoreError::AttributeError { span, .. } => *span,
            EsFluentCoreError::VariantError { span, .. } => *span,
            EsFluentCoreError::FieldError { span, .. } => *span,
            EsFluentCoreError::TransformError { span, .. } => *span,
        }
    }

    pub fn message_mut(&mut self) -> &mut String {
        match self {
            EsFluentCoreError::AttributeError { message, .. } => message,
            EsFluentCoreError::VariantError { message, .. } => message,
            EsFluentCoreError::FieldError { message, .. } => message,
            EsFluentCoreError::TransformError { message, .. } => message,
        }
    }

    pub fn abort(self) -> ! {
        let msg = self.to_string();
        match self.span() {
            Some(span) => abort!(span, "{}", msg),
            None => abort_call_site!("{}", msg),
        }
    }

    pub fn emit(&self) {
        let msg = self.to_string();
        match self.span() {
            Some(span) => emit_error!(span, "{}", msg),
            None => emit_error!("{}", msg),
        }
    }
}

pub trait ErrorExt {
    fn with_note(self, note: String) -> Self;
    fn with_help(self, help: String) -> Self;
}

impl ErrorExt for EsFluentCoreError {
    fn with_note(mut self, note_msg: String) -> Self {
        let message = self.message_mut();
        *message = format!("{}\nnote: {}", message, note_msg);
        self
    }

    fn with_help(mut self, help_msg: String) -> Self {
        let message = self.message_mut();
        *message = format!("{}\nhelp: {}", message, help_msg);
        self
    }
}

pub type EsFluentCoreResult<T> = Result<T, crate::error::EsFluentCoreError>;
