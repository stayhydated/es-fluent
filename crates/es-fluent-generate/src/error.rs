use thiserror::Error;

#[derive(Debug, Error)]
pub enum FluentGenerateError {
    /// An IO error.
    #[error("IO error: {0}")]
    Io(#[from] std::io::Error),

    /// An environment variable error.
    #[error("Environment variable error: {0}")]
    EnvVar(#[from] std::env::VarError),

    /// An error that occurs when the package name is missing.
    #[error("Missing package name")]
    MissingPackageName,

    /// An error that occurs when parsing a Fluent file.
    #[error("Fluent parsing error: {0:?}")]
    ParseError(Vec<fluent_syntax::parser::ParserError>),

    /// An error that occurs when serializing a Fluent file.
    #[error("Fluent serialization error: {0}")]
    SerializeError(#[from] std::fmt::Error),
}
