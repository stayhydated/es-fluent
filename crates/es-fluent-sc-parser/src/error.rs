use std::path::PathBuf;
use thiserror::Error;

#[derive(Debug, Error)]
pub enum FluentScParserError {
    #[error("IO error accessing path '{0}': {1}")]
    Io(PathBuf, #[source] std::io::Error),

    #[error("Failed to parse Rust file '{0}': {1}")]
    Syn(PathBuf, #[source] syn::Error),

    #[error("Error walking directory '{0}': {1}")]
    WalkDir(PathBuf, #[source] walkdir::Error),

    #[error("Attribute parsing error in file '{0}': {1}")]
    AttributeParse(PathBuf, #[source] darling::Error),

    #[error("Missing required attribute data in file '{0}': {1}")]
    MissingAttribute(PathBuf, String),

    #[error("Internal logic error: {0}")]
    Internal(String),
}
