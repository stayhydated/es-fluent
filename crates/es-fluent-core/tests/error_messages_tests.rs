use es_fluent_core::error::{ErrorExt, EsFluentCoreError};
use proc_macro2::Span;

#[test]
fn attribute_error_message_snapshot() {
    let err = EsFluentCoreError::AttributeError {
        message: "Unknown attribute 'foo'".to_string(),
        span: None,
    };

    insta::assert_ron_snapshot!(
        "error_messages__attribute_error_message_snapshot",
        err.to_string()
    );
}

#[test]
fn variant_error_with_note_and_help_snapshot() {
    let err = EsFluentCoreError::VariantError {
        message: "Unsupported variant style".to_string(),
        variant_name: "Data".to_string(),
        span: Some(Span::call_site()),
    }
    .with_note("Only unit or tuple variants are supported here".to_string())
    .with_help("Consider splitting 'Data' into unit + struct variants as needed".to_string());

    insta::assert_ron_snapshot!(
        "error_messages__variant_error_with_note_and_help_snapshot",
        err.to_string()
    );
}

#[test]
fn field_error_named_and_unnamed_snapshot() {
    let named = EsFluentCoreError::FieldError {
        message: "Cannot be both #[fluent(skip)] and #[fluent(default)]".to_string(),
        field_name: Some("value".to_string()),
        span: None,
    };

    let unnamed = EsFluentCoreError::FieldError {
        message: "Unexpected attribute on tuple field".to_string(),
        field_name: None,
        span: None,
    };

    let cases = vec![
        ("named", named.to_string()),
        ("unnamed", unnamed.to_string()),
    ];

    insta::assert_ron_snapshot!(
        "error_messages__field_error_named_and_unnamed_snapshot",
        &cases
    );
}

#[test]
fn transform_error_message_snapshot() {
    let err = EsFluentCoreError::TransformError {
        message: "Failed to convert type: Path<'a> -> String".to_string(),
        span: None,
    }
    .with_help("Implement TryFrom<Path<'a>> for String or adjust your types".to_string());

    insta::assert_ron_snapshot!(
        "error_messages__transform_error_message_snapshot",
        err.to_string()
    );
}
