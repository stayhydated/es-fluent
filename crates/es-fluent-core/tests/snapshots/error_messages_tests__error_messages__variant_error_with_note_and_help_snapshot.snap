---
source: crates/es-fluent-core/tests/error_messages_tests.rs
expression: err.to_string()
---
"Variant \'Data\' error: Unsupported variant style\nnote: Only unit or tuple variants are supported here\nhelp: Consider splitting \'Data\' into unit + struct variants as needed"
